src/routing_table/prefix.rs

```diff
@@ -213,83 +282,99 @@ impl Prefix {
+impl PartialOrd<Self> for UnversionedPrefix {
     fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
         Some(self.cmp(other))
     }
 }
 
-impl Ord for Prefix {
+impl Ord for UnversionedPrefix {
     fn cmp(&self, other: &Self) -> Ordering {
-        if self == other {
-            Ordering::Equal
-        } else if self.is_compatible(other) {
+        if self.is_compatible(other) {
             self.bit_count().cmp(&other.bit_count())
         } else {
             self.name.cmp(&other.name)
         }
     }
 }
 
-impl Hash for Prefix {
+impl Hash for UnversionedPrefix {
     fn hash<H: Hasher>(&self, state: &mut H) {
         for i in 0..self.bit_count() {
             self.name.bit(i).hash(state);
         }
     }
 }
 
-impl Binary for Prefix {
+impl Binary for UnversionedPrefix {
     fn fmt(&self, formatter: &mut Formatter) -> FmtResult {
         let mut binary = self.name.binary();
         binary.truncate(self.bit_count());
         write!(formatter, "{}", binary)
     }
 }
 
-impl Debug for Prefix {
+impl Debug for UnversionedPrefix {
     fn fmt(&self, formatter: &mut Formatter) -> FmtResult {
         write!(formatter, "Prefix({:b})", self)
     }
 }
 
-#[cfg(test)]
-impl FromStr for Prefix {
-    type Err = String;
-    fn from_str(bits: &str) -> Result<Prefix, String> {
-        let mut first_byte = 0u8;
-        for (i, bit) in bits.chars().enumerate() {
-            if bit == '1' {
-                first_byte |= 1 << (7 - i);
-            } else if bit != '0' {
-                return Err(format!(
-                    "'{}' not allowed - the string must represent a binary number.",
-                    bit
-                ));
-            }
+/// Find the entry at the given prefix, ignoring versions.
+pub fn unversioned_find<'a, T>(
+    map: &'a BTreeMap<Prefix, T>,
+    key: &UnversionedPrefix,
+) -> Option<(Prefix, &'a T)> {
+    map.range(key.with_version(0)..key.with_version(u64::MAX))
+        .next()
+        .map(|(prefix, value)| (*prefix, value))
+}
+
+/// Get the value at the given prefix, ignoring versions.
+pub fn unversioned_get<'a, T>(
+    map: &'a BTreeMap<Prefix, T>,
+    key: &UnversionedPrefix,
+) -> Option<&'a T> {
+    map.range(key.with_version(0)..key.with_version(u64::MAX))
+        .next()
+        .map(|(_, value)| value)
+}
+
+/// Check whether the map contains a key equal to the given prefix but ignoring versions.
+pub fn unversioned_contains_key<'a, T>(
+    map: &'a BTreeMap<Prefix, T>,
+    key: &UnversionedPrefix,
+) -> bool {
+    map.range(key.with_version(0)..key.with_version(u64::MAX))
+        .next()
+        .is_some()
+}
+
+/// Remove and return the element at the given prefix, ignoring versions.
+pub fn unversioned_remove(set: &mut BTreeSet<Prefix>, key: &UnversionedPrefix) -> Option<Prefix> {
+    if let Some(prefix) = set.range(key.with_version(0)..key.with_version(u64::MAX))
+        .cloned()
+        .next()
+    {
+        if set.remove(&prefix) {
```

Comment:
it looks like when we carry out the removal, the version does matter?
i.e. if the entry returned doesn't have the matched version, it won't get removed?

Reply:
No, it first finds the (versioned) `Prefix` matching the given `UnversionedPrefix` (via `range`). Then it removes it.

Reply:
Really nice way of using iterators 👍  I'd probably suggest even adding a helper method to replace `.range(pfx.with_version(0)..key.with_version(u64::MAX))` with something like `.all_versions_range(pfx)`, with an impl like
```
impl Iter<Item = T> {
  ...
  fn all_versions_range(&self, pfx: Prefix) {
    self.range(pfx.with_version(0)..pfx.with_version(u64::MAX))
  }
}
```
This might convey the intent in a more clean manner IMO. Something to note for future refactorings.